### PR TITLE
Sprint 21

### PR DESCRIPTION
**Pull request by urweakness**
Организовал передачу зависимостей и контроль состояния загрузки через Coordinator/DI с изолированным контекстом выполнения + загрузки по таску ( .task{} ) во вью моделях [async await].

API кривой, я реализовал свою часть так, как оно должно работать, пересадки не возвращаются не по моей вине. Есть хардкод метод принудительной конкретной вставки даты на которую приходит респонс только с пересадками, где нужно делать запрос по городу вместо станции, где респонс приходит по пагинации, а это сплошные компромиссы, и это полностью противоречит всем законам логики, поэтому прошу принять именно этот вариант

### CODE REVIEW

**Comment by urweakness** [comment]
в сторисах под текст положил мягкий бекграунд (нет в дизайне, но это бред, текст смешивается с фоткой, читать не возможно, на реальном проекте такого никогда не будет)

File: TravelSchedule/Presentation/Carriers/CarrierListCellView.swift [hunk 1]
```diff
@@ -1,43 +1,74 @@
...
-            .frame(height: 104)
-            .task {
-                Task {
-					await loadCarrierImage()
-                }
-            }
+			}
+		}
+		.padding(.top, 14)
+		.padding(.leading, 14)
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/Carriers/CarrierListCellView.swift:56]: Можно объединить в один модификатор:
`.padding([.top, .leading], 14)`

File: TravelSchedule/Presentation/Carriers/CarrierListCellView.swift [hunk 2]
```diff
@@ -107,38 +140,16 @@ struct CarrierListCellView: View {
...
+    // --- private methods ---
 	private func loadCarrierImage() async {
+		guard let logoURLString = carrier.logoURLString else { return }
+		
 		let loader = DataLoader()
-		guard let imageURL = URL(string: carrier.logoURLString) else { return }
+		guard let imageURL = URL(string: logoURLString) else { return }
 		guard
-			let data = await loader.downloadData(url: imageURL),
+			let data = try? await loader.downloadData(url: imageURL),
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/Carriers/CarrierListCellView.swift]: Можно объединить в один блок `guard`

File: TravelSchedule/Presentation/Carriers/CarrierParser.swift [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		var result = [CarrierModel]()
+		for segment in response.segments ?? [] {
+			guard
+				let carrier = parseSegment(segment: segment)
+			else { continue }
+			
+			result.append(carrier)
+		}
+		return result
+	}
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/Carriers/CarrierParser.swift:10]: Можно упростить:
``` swift
func parseResponse(_ response: ScheduleBetweenStationsResponse) -> [CarrierModel] {
   (response.segments ?? [])
      .compactMap(parseSegment)
}
```

File: TravelSchedule/Presentation/Carriers/CarriersListView.swift [hunk 4]
```diff
@@ -1,67 +1,64 @@
...
-                                }
-                        }
-                    }
-                    .padding(.bottom, overscrollBottomPadding)
-                }
-                .scrollIndicators(.hidden)
+				if loadingState == .idle {
+					carriersListView
+						.transition(.opacity)
+				} else {
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/Carriers/CarriersListView.swift]: Везде, где это возможно (если layout при этом не изменяется), лучше использовать `opacity` вместо `if` для сокрытия `View`. [Вот хороший пример с ссылками на `WWDC`](https://stackoverflow.com/a/57420479)

File: TravelSchedule/Presentation/Carriers/CarriersListView.swift [hunk 5]
```diff
@@ -70,33 +67,79 @@ struct CarriersListView: View {
...
     private func loadCarriers() async {
-        #warning("TODO: load carriers using (travelRoutingManager.filter)")
+		guard
+			let response = await viewModel.fetchScheduleBetweenStations()
+		else {
+			print("\n\(#function) No response from API\n")
+			return
+		}
+		
+		await viewModel.parseResponse(response)
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/Carriers/CarriersListView.swift]: Можно исправить предупреждение: 
`No 'async' operations occur within 'await' expression`

File: TravelSchedule/Presentation/SettingsView/WebViewRepresentable.swift [hunk 6]
```diff
@@ -29,45 +32,114 @@ struct WebViewRepresentable: UIViewRepresentable {
...
 	func makeCoordinator() -> Coordinator {
-		Coordinator(isLoading: $isLoading)
+		Coordinator(
+			isLoading: $isLoading,
+			onProgress: onProgress
+		)
 	}
 	
-	class Coordinator: NSObject {
+	class Coordinator: NSObject, WKNavigationDelegate {
```
Comment (NaR0RoW) [inline TravelSchedule/Presentation/SettingsView/WebViewRepresentable.swift:41]: Стоит отметить `final`

File: TravelSchedule/Services/CarrierService.swift [hunk 7]
```diff
@@ -20,6 +25,28 @@ final class CarrierService: CarrierServiceProtocol {
...
         )
         
-        return try response.ok.body.json
+		switch response {
+		case .ok(let content):
+			return .success(try content.body.json)
+		case .undocumented(let statusCode, let description):
+			switch statusCode {
+			case 400...499:
+				return .failure(.noInternet)
```
Comment (NaR0RoW) [inline TravelSchedule/Services/CarrierService.swift:34]: 400-499 ≠ нет интернета - это клиентские ошибки (неправильный запрос, неавторизован, не найдено и т.д.)

File: TravelSchedule/Services/CopyrightService.swift [hunk 8]
```diff
@@ -1,21 +1,48 @@
...
+							domain: "StationListService",
+							code: 0,
+							userInfo: ["description": description])
+					)
+				)
+			}
+			
+		@unknown default:
+			throw URLError(.badServerResponse)
+		}
```
Comment (NaR0RoW) [inline TravelSchedule/Services/CopyrightService.swift:46]: Обработка response в акторах практически одинакова. Лучше избегать дублирования и максимально переиспользовать код

File: TravelSchedule/Services/StationsListService.swift [hunk 9]
```diff
@@ -1,46 +1,73 @@
...
+					return .success(payload)
+					
+				case .html(let body):
+					
+					var data = Data()
+					for try await chunk in body {
+						data.append(contentsOf: chunk)
+					}
+					
+					let response = try JSONDecoder().decode(AllStationsResponse.self, from: data)
```
Comment (NaR0RoW) [inline TravelSchedule/Services/StationsListService.swift]: `JSONDecoder` - это тяжеловесный объект. Нет смысла создавать его каждый раз внутри запроса. Лучше создать один экземпляр как свойство сервиса и переиспользовать - это экономит ресурсы и улучшает читаемость кода

File: TravelSchedule/ViewModels/StoriesViewModel.swift [hunk 10]
```diff
@@ -1,55 +1,135 @@
...
+		var result: CGFloat = 0
+		for i in 0..<count {
+			if i < partIndex {
+				result += 1
+			} else if i == partIndex {
+				result += partProgress
+			}
+		}
+		return result / CGFloat(count)
+	}
```
Comment (NaR0RoW) [inline TravelSchedule/ViewModels/StoriesViewModel.swift:120]: Кажется можно следующим образом записать:
``` swift
func progressForCurrentStoryProgress(for story: StoryModel) -> CGFloat {
    let totalParts = max(1, story.storyParts.count)
    let completedProgress = CGFloat(min(currentStoryPartIndex, totalParts)) + timer.progress
    return completedProgress / CGFloat(totalParts)
}
```

File: TravelSchedule/ViewModels/TravelPointChooseViewModel.swift [hunk 11]
```diff
@@ -1,40 +1,54 @@
...
             .receive(on: DispatchQueue.main)
             .assign(to: \.filteredObjects, on: self)
             .store(in: &cancellables)
     }
 
-    private func filterObjects(for searchText: String) -> [D] {
+    private func parseObjects(for searchText: String) -> [D] {
         let clearText = searchText.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
 
         if clearText.isEmpty {
```
Comment (NaR0RoW) [inline TravelSchedule/ViewModels/TravelPointChooseViewModel.swift]: Можно упростить код, используя тернарный оператор:
``` swift
let clearText = searchText.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
return clearText.isEmpty ? Array(travelPoints) : travelPoints.filter {
   $0.name.lowercased().contains(clearText)
}
```

File: TravelSchedule/ViewModels/TravelPointChooseViewModel.swift [hunk 12]
```diff
@@ -45,4 +59,119 @@ final class TravelPointChooseViewModel<D: TravelPoint>: ObservableObject {
...
+					Town(
+						name: title,
+						code: code
+					)
+				)
+			}
+		}
+		
+		return result
+	}
```
Comment (NaR0RoW) [inline TravelSchedule/ViewModels/TravelPointChooseViewModel.swift]:
``` swift
private func parseTowns() -> [Town] {
    guard let countries = appManager.stationList?.countries,
          let russia = countries.first(where: { $0.title == "Россия" }),
          let regions = russia.regions else {
        return []
    }
    return regions
        .compactMap { $0.settlements }
        .flatMap { $0 }
        .compactMap { settlement in
            guard let title = settlement.title,
                  let code = settlement.codes?.yandex_code else {
                return nil
            }
            return Town(name: title, code: code)
        }
}
```

File: TravelSchedule/ViewModels/TravelPointChooseViewModel.swift [hunk 13]
```diff
@@ -45,4 +59,119 @@ final class TravelPointChooseViewModel<D: TravelPoint>: ObservableObject {
...
+								yandexCode: station.codes?.yandex_code
+							)
+						)
+					}
+				}
+			}
+		}
+		
+		return result
+	}
```
Comment (NaR0RoW) [inline TravelSchedule/ViewModels/TravelPointChooseViewModel.swift]: Просто предлагаю более компактные варианты:
``` swift
private func parseStations() -> [Station] {
    guard let isDestination = routingManager.isDestination,
          let targetTownName = isDestination ? routingManager.destinationTown?.name : routingManager.startTown?.name,
          let countries = appManager.stationList?.countries else {
        return []
    }
    return countries
        .flatMap { $0.regions ?? [] }
        .flatMap { $0.settlements ?? [] }
        .filter { $0.title == targetTownName }
        .flatMap { $0.stations ?? [] }
        .compactMap { station in
            guard let title = station.title else { return nil }
            return Station(
                name: title,
                esrCode: station.codes?.esr_code,
                yandexCode: station.codes?.yandex_code
            )
        }
}
```

**Review by NaR0RoW — APPROVED** [review]
Отличная работа, код хорошо написан 🔥
